projects/common-api/src/main/java/dan200/computercraft/api/upgrades/UpgradeBase.java

```diff
@@ -50,6 +52,45 @@ public interface UpgradeBase {
      */
     ItemStack getCraftingItem();
 
+    /**
+     * Return an item stack representing the instance of upgrade with passed upgrade data.
+     * By default, this method output equals to {@link #getCraftingItem()} because default behavior of
+     * upgrades is ignored any upgrade data.
+     * <p>
+     * But if your upgrade required this, you can redefine this behavior and pass some information from upgrade data
+     * to item stack. Please, use this method with cautions.
+     * <p>
+     * This method will be used by {@code turtle.unequipLeft()} and {@code pocket.unequipBack()} to determine item stack
+     * that should be placed inside inventory.
+     *
+     * @param upgradeData NBT information that was stored inside turtle/pocket by upgrade
+     * @return The item stack, that should be stored inside inventory
+     */
+    default @Nonnull ItemStack getUpgradeItem(@Nonnull CompoundTag upgradeData) {
+        return getCraftingItem();
+    }
+
+    /**
+     * Returns initial upgrade data, that should be set for turtle/pocket, that uses this upgrade, based on item stack
+     * that used for installing upgrade right now.
+     * @param stack Item Stack that used for upgrade
+     * @return Upgrade NBT data that should be set in time of adding this upgrade to turtle/pocket
+     */
+    default @Nonnull CompoundTag getUpgradeData(@Nonnull ItemStack stack) {
+        return new CompoundTag();
+    }
+
+    /**
+     * Specific hook, that allow you to filter data, that should be stored when turtle was broken.
+     * Use this in cases when you don't need to store all upgrade data by default, and you want to change this
+     * behavior, for example, as modem do.
+     * @param upgradeData NBT data that currently stored for this upgrade
+     * @return filtered version of this data, by default just all of it
+     */
+    default @Nonnull CompoundTag getPersistedData(CompoundTag upgradeData) {
+        return upgradeData;
```

Comment:
Did we want to make this return a new compound tag? Happy to go with either, just wasn't sure what the conclusion of our discussion was.

Reply:
So, your initial suggestion was to return same tag. At first, I was thinking about returning empty new Compoung Tag, but I think now that returning same tag is much intuitive way to work this.

Reply:
I have updated ModemUpgrade code to return empty CompundTag, so it will not store modem state on turtle break